quil-rs/src/program/calibration.rs

```diff
@@ -203,6 +199,28 @@ impl CalibrationSet {
         })
     }
 
+    /// Returns the last-specified [`MeasureCalibrationDefinition`] that matches the target
+    /// qubit (if any), or otherwise the last-specified one that specified no qubit.
+    pub fn get_match_for_measurement(
+        &self,
+        measurement: &Measurement,
+    ) -> Option<&MeasureCalibrationDefinition> {
+        let mut matching_calibration = None;
+        let mut found_matching_calibration_without_qubit = false;
+        for cal in self.measure_calibrations.iter().rev() {
+            if let Some(cal_qubit) = &cal.qubit {
+                if cal_qubit == &measurement.qubit || matches!(cal_qubit, Qubit::Variable(_)) {
```

Comment:
@kalzoo @notmgsk 

This was refactored out of expand, the only addition is the addition of ` || matches!(cal_qubit, Qubit::Variable(_))`. I added this so these instructions like these would match:

```
"DEFCAL MEASURE q dest", "MEASURE 0"
```

Is this correct?

Reply:
Not quite. The fact that that matches is right (except for the lack of mref in the MEASURE, I think that should be a non-match or not allowed at all) - but the bug is in the `break`. There could be a closer match later on, much like the qubit matching for a gate. That's what `found_matching_calibration_without_qubit = true` is doing - it has a matching calibration but then keeps on looking for a later match. The order of precedence:

1. Match fixed qubit
2. Match variable qubit
3. Match no qubit (I don't recall this case but it's supported here)

The tiebreaker among calibrations of equal precedence is "last wins."

That being said, it looks like there are other pre-existing bugs in this implementation which don't follow the "last wins" rule. So your change just extends an existing bug - this could be fixed now or punted.

Reply:
Cool, I've started a fix, but it requires a bit more thought and would detract from the goal of this PR, so I created #160 to track.

Reply:
I've also made it so that a `MemoryReference` is required to match a `Measurement`.